src/Common/Core/Impl/Threading/BinaryAsyncLock.cs

```diff
@@ -11,39 +11,95 @@ namespace Microsoft.Common.Core.Threading {
     /// All other callers will wait until <see cref="Release"/> is called and then will get <see langword="true" />
     /// </summary>
     public class BinaryAsyncLock {
-        private TaskCompletionSource<bool> _tcs;
+        private static readonly Task<bool> FalseTask = Task.FromResult(false);
+        private static readonly Task<bool> TrueTask = Task.FromResult(true);
+        private readonly Queue<TaskCompletionSource<bool>> _queue = new Queue<TaskCompletionSource<bool>>();
```

Comment:
Too much has to be done as an atomic operation. The main problem is synchronizing `_isCompleted` with `_queue` operations. For example, concurrent calls of `Release` and `Reset` can bring the lock into state when it is considered released, but some of the tasks from queue are completed with `false` result

---
In reply to: [78201950](https://github.com/Microsoft/RTVS/pull/2268#discussion_r78201950) [](ancestors = 78201950)